File: specodex/relations.py
"""Device-relations layer for cross-product compatibility queries.

Powers `/api/v1/relations/*` endpoints (planned, separate PR) and admin
CLI flows that ask "which motors fit this actuator", "which drives can
run this motor", "which gearheads accept this motor frame".

The compatibility predicates here read Pydantic model fields directly,
so schema drift surfaces at type-check time. See `todo/SCHEMA.md` Part 3
for the design rationale and Phase 1 for the bridge fields
(`motor_mount_pattern`, `compatible_motor_mounts`,
`input_motor_mount`) that make these queries typed instead of
string-comparison hacks.

Predicate philosophy: **exclude on missing data, not include**. If a
candidate record lacks a field the predicate needs, it is omitted from
the result rather than passed through. The compatibility query is meant
to be precise — "compatible motors" should never include "we don't
know if this is compatible" rows. Recall loss from incomplete records
is the right failure mode here; precision loss would put the wrong
hardware in a Build BOM.
"""

from __future__ import annotations

from typing import Iterable, List, Optional, Union

from specodex.models.common import MinMaxUnit, ValueUnit
from specodex.models.drive import Drive
from specodex.models.electric_cylinder import ElectricCylinder
from specodex.models.gearhead import Gearhead
from specodex.models.linear_actuator import LinearActuator
from specodex.models.motor import Motor


ActuatorLike = Union[LinearActuator, ElectricCylinder]


# ---------------------------------------------------------------------------
# Internal predicates. All return False on any None / missing-unit input
# per the "exclude on missing data" rule above.
# ---------------------------------------------------------------------------


def _value_in_range(value: Optional[ValueUnit], rng: Optional[MinMaxUnit]) -> bool:
    """True if `value` falls within `rng` [min, max] (unit-aware).

    Both endpoints of `rng` are optional: an open-ended range
    (min=None, max=120) means "anything ≤ 120 in unit"; (min=24, max=None)
    means "≥ 24 in unit". Units must match — the Pydantic
    BeforeValidators on VoltageRange / Current normalise to canonical
    units on construction, so a unit mismatch here is a genuine
    incompatibility, not a representation bug.
    """
    if value is None or rng is None:
        return False
    if value.value is None or value.unit is None or rng.unit is None:
        return False
    if value.unit != rng.unit:
        return False
    if rng.min is not None and value.value < rng.min:
        return False
    if rng.max is not None and value.value > rng.max:
        return False
    return True


def _range_within(inner: Optional[MinMaxUnit], outer: Optional[MinMaxUnit]) -> bool:
    """True if `inner` is fully contained in `outer` (unit-aware).

    Used for voltage matching: a motor rated for 200-240V must run on a
    drive whose input range covers that whole span. A drive with input
    480-480V doesn't cover a 200-240V motor even though the motor's
    high end is below the drive's low end.
    """
    if inner is None or outer is None:
        return False
    if inner.unit is None or outer.unit is None or inner.unit != outer.unit:
        return False
    # If `inner` has no min, treat as -inf for the lower-bound check; same
    # for max. This keeps single-point voltages (min==max) and one-sided
    # ranges from short-circuiting incorrectly.
    if inner.min is not None:
        if outer.min is not None and inner.min < outer.min:
            return False
    if inner.max is not None:
        if outer.max is not None and inner.max > outer.max:
            return False
    # Both inner endpoints None → there is no voltage info to check; the
    # outer dictates nothing. Conservative: reject as missing-data.
    if inner.min is None and inner.max is None:
        return False
    return True


def _value_gte(a: Optional[ValueUnit], b: Optional[ValueUnit]) -> bool:
    """True if `a >= b` (unit-aware). Any None / unit mismatch returns False."""
    if a is None or b is None:
        return False
    if a.value is None or b.value is None:
        return False
    if a.unit is None or b.unit is None or a.unit != b.unit:
        return False
    return a.value >= b.value


def _shaft_compatible(
    motor_shaft: Optional[ValueUnit], gearhead_input: Optional[ValueUnit]
) -> bool:
    """Motor shaft fits if its diameter is at most the gearhead's input bore.

    ``input_shaft_diameter`` on a gearhead is the MAXIMUM motor shaft the
    input coupling accepts — precision-gearbox catalogs publish it as a
    clamping-bushing bound ("motor shaft ≤ 24 mm"), and bushing kits
    cover smaller shafts. So the check is `motor ≤ bore`, not equality.

    The 0.1mm grace absorbs catalog rounding ("10mm" vs "9.95mm") on the
    boundary; final fit is checked against actual mechanical drawings,
    not this predicate.
    """
    if motor_shaft is None or gearhead_input is None:
        return False
    if motor_shaft.value is None or gearhead_input.value is None:
        return False
    if motor_shaft.unit is None or gearhead_input.unit is None:
        return False
    if motor_shaft.unit != gearhead_input.unit:
        return False
    return motor_shaft.value <= gearhead_input.value + 0.1


def _meets_floor(value: Optional[ValueUnit], floor: float, unit: str) -> bool:
    """True if `value` is present, in canonical `unit`, and >= `floor`.

    Compares an actuator spec field against a numeric requirement floor
    already expressed in canonical units (the Build derivation engine
    emits mm / N / mm·s⁻¹). Typed family fields (`stroke`,
    `max_push_force`) normalise aliases to canonical on construction, so
    a unit other than `unit` means either a different physical quantity
    or an un-normalised generic field — excluded either way, per this
    module's precision-first rule.
    """
    if value is None or value.value is None or value.unit is None:
        return False
    if value.unit != unit:
        return False
    return value.value >= floor


def input_torque_capacity(g: Gearhead) -> Optional[ValueUnit]:
    """Continuous input-torque capacity of a gearhead, referred from output.

    Catalogs rate output torque (T2N); the motor cares about the input
    side. T_in = T2N / (ratio × efficiency). Efficiency defaults to 1.0
    when absent or out of (0, 1] — that UNDERSTATES capacity (real
    losses mean the input can actually push a bit more before the
    output hits T2N), so missing data errs toward excluding, per this
    module's precision-first rule.

    Returns None when `gear_ratio` or a positive `max_continuous_torque`
    is missing — callers exclude the row.
    """
    tq = g.max_continuous_torque
    if tq is None or tq.value is None or tq.unit is None:
        return None
    if g.gear_ratio is None or g.gear_ratio <= 0:
        return None
    eff = g.efficiency if g.efficiency is not None and 0 < g.efficiency <= 1 else 1.0
    return ValueUnit(value=tq.value / (g.gear_ratio * eff), unit=tq.unit)


def _gearhead_speed_ok(motor: Motor, g: Gearhead) -> bool:
    """Motor's rated speed must sit within the gearhead's input-speed rating.

    Prefers `max_input_speed` (n1B — the hard ceiling); falls back to
    `nominal_input_speed` (n1N — stricter, continuous rating) when the
    ceiling wasn't published. Missing both → excluded.
    """
    rating = (
        g.max_input_speed if g.max_input_speed is not None else g.nominal_input_speed
    )
    return _value_gte(rating, motor.rated_speed)


def _encoder_protocol_intersect(motor: Motor, drive: Drive) -> bool:
    """True if motor's encoder protocol is in drive's supported list.

    Schema is mid-harmonisation (todo/SCHEMA.md Phase 1.1, also touched
    by the in-flight DOUBLE_TAP work): on master, motor side is
    ``Optional[str]`` and drive side is ``Optional[List[str]]``. The
    DOUBLE_TAP branch promotes motor to a structured ``EncoderFeedback``
    with a ``.protocol`` attribute. Tolerate both shapes here so this
    predicate doesn't have to change when DOUBLE_TAP merges.

    Drive side is always a list (or None) — easy to test membership
    against. Motor side is the string OR the protocol attribute on the
    structured model.
    """
    motor_enc = motor.encoder_feedback_support
    drive_protocols = drive.encoder_feedback_support
    if motor_enc is None or drive_protocols is None:
        return False
    if isinstance(motor_enc, str):
        motor_proto: Optional[str] = motor_enc
    else:
        motor_proto = getattr(motor_enc, "protocol", None)
    if motor_proto is None:
        return False
    return motor_proto in drive_protocols


# ---------------------------------------------------------------------------
# Public compatibility queries.
# ---------------------------------------------------------------------------


def compatible_actuators(
    actuator_db: Iterable[LinearActuator],
    *,
    min_stroke_mm: Optional[float] = None,
    min_peak_force_n: Optional[float] = None,
    min_peak_velocity_mm_s: Optional[float] = None,
) -> List[LinearActuator]:
    """Linear actuators meeting the requirement floors Build derives.

    The requirements-first entry point for Build's slot-fill sequence
    (`todo/BUILD.md` Part 4): given the numeric constraints Build's
    derivation engine produces from a motion application — minimum
    stroke, peak thrust, peak velocity — return the actuators whose
    ratings clear every supplied floor.

    Each floor is optional and independent. A `None` floor applies no
    constraint on that dimension (Build's "blank = no constraint
    applied" rule). A floor that *is* set excludes any actuator missing
    the corresponding field or carrying it in a non-canonical unit —
    precision over recall, consistent with the rest of this module:
    "compatible" must never include "we can't tell".

    Field mapping:
    - `min_stroke_mm`          → `stroke` (Length, canonical mm)
    - `min_peak_force_n`       → `max_push_force` (Force, canonical N —
      the carriage thrust rating)
    - `min_peak_velocity_mm_s` → `max_linear_speed` (generic mm/s)

    Orientation and duty cycle from the Build requirements are
    deliberately NOT actuator filters: orientation is a derating hint
    the downstream motor sizing consumes, and duty cycle is a
    motor-thermal concept with no actuator-rating analogue. They belong
    to the slots that can act on them, not here.
    """
    out: List[LinearActuator] = []
    for a in actuator_db:
        if min_stroke_mm is not None and not _meets_floor(
            a.stroke, min_stroke_mm, "mm"
        ):
            continue
        if min_peak_force_n is not None and not _meets_floor(
            a.max_push_force, min_peak_force_n, "N"
        ):
            continue
        if min_peak_velocity_mm_s is not None and not _meets_floor(
            a.max_linear_speed, min_peak_velocity_mm_s, "mm/s"
        ):
            continue
        out.append(a)
    return out


def compatible_motors(
    actuator: ActuatorLike,
    motor_db: Iterable[Motor],
    *,
    min_torque: Optional[ValueUnit] = None,
    min_speed: Optional[ValueUnit] = None,
) -> List[Motor]:
    """Motors that mount on `actuator` and meet optional torque/speed floor.

    Mount matching:
    - LinearActuator: motor's frame must be in `compatible_motor_mounts`
      (a list — actuators often accept multiple frames via adapter plates).
    - ElectricCylinder: motor's frame must equal `motor_mount_pattern`
      (single-valued — cylinders are typically one integrated package).

    Returns empty list if the actuator has no mount info on file. The
    compatibility query is precise; "show every motor" is the wrong
    fallback when the actuator's mount pattern hasn't been ingested.
    """
    if isinstance(actuator, LinearActuator):
        mounts = set(actuator.compatible_motor_mounts or [])
    else:  # ElectricCylinder
        mounts = (
            {actuator.motor_mount_pattern} if actuator.motor_mount_pattern else set()
        )
    if not mounts:
        return []

    out: List[Motor] = []
    for m in motor_db:
        if m.motor_mount_pattern is None or m.motor_mount_pattern not in mounts:
            continue
        if min_torque is not None and not _value_gte(m.rated_torque, min_torque):
            continue
        if min_speed is not None and not _value_gte(m.rated_speed, min_speed):
            continue
        out.append(m)
    return out


def compatible_drives(motor: Motor, drive_db: Iterable[Drive]) -> List[Drive]:
    """Drives whose voltage / current envelope covers this motor.

    A drive is compatible if all three hold:
    1. Drive's `input_voltage` range fully contains motor's `rated_voltage`
       range. Single-point motor voltages (min == max) are handled the
       same way — the inner range collapses to a point that must be in
       the outer.
    2. Drive's `rated_current` is at least the motor's `rated_current`
       (the drive can deliver continuous current the motor demands).
    3. Motor's encoder protocol appears in drive's supported list (the
       wire format matches). Devices behind the wire are the motor's
       problem, not the drive's.

    Returns empty list if motor lacks any of the three required fields.
    """
    if motor.rated_voltage is None or motor.rated_current is None:
        return []

    out: List[Drive] = []
    for d in drive_db:
        if not _range_within(motor.rated_voltage, d.input_voltage):
            continue
        if not _value_gte(d.rated_current, motor.rated_current):
            continue
        if not _encoder_protocol_intersect(motor, d):
            continue
        out.append(d)
    return out


def compatible_gearheads(
    motor: Motor, gearhead_db: Iterable[Gearhead]
) -> List[Gearhead]:
    """Gearheads whose input side accepts this motor as an ideal match.

    A gearhead is compatible if all four hold:
    1. Mount does not CONTRADICT: when both sides publish mount info,
       the gearhead's `input_motor_mount` list must contain the motor's
       `motor_mount_pattern`. When either side lacks it, the mount
       check is skipped — precision-gearbox vendors sell adapter
       plates instead of publishing NEMA-style patterns (only ~5% of
       ingested gearheads carry a list, 2026-07-25), so treating a
       missing list as a mismatch would exclude nearly the whole
       catalog on a field the vendor never states. This is a
       deliberate, documented exception to the module's
       exclude-on-missing rule; the physical checks below stay hard.
    2. The motor's `shaft_diameter` fits within its
       `input_shaft_diameter` (max-bore semantics, 0.1mm rounding
       grace — see `_shaft_compatible`).
    3. The motor's `rated_speed` sits within the gearhead's input-speed
       rating (`max_input_speed`, falling back to
       `nominal_input_speed` — see `_gearhead_speed_ok`).
    4. The motor's `rated_torque` does not exceed the gearhead's
       continuous input-torque capacity, referred from output torque
       through ratio and efficiency (see `input_torque_capacity`).

    Returns empty list if the motor lacks any of `shaft_diameter`,
    `rated_speed`, or `rated_torque` — mirroring `compatible_drives`'
    treatment of required motor fields.
    """
    if (
        motor.shaft_diameter is None
        or motor.rated_speed is None
        or motor.rated_torque is None
    ):
        return []

    out: List[Gearhead] = []
    for g in gearhead_db:
        if (
            motor.motor_mount_pattern is not None
            and g.input_motor_mount
            and motor.motor_mount_pattern not in g.input_motor_mount
        ):
            continue
        if not _shaft_compatible(motor.shaft_diameter, g.input_shaft_diameter):
            continue
        if not _gearhead_speed_ok(motor, g):
            continue
        if not _value_gte(input_torque_capacity(g), motor.rated_torque):
            continue
        out.append(g)
    return out


def _fmt(v: Optional[ValueUnit]) -> str:
    if v is None or v.value is None:
        return "?"
    return f"{v.value:g} {v.unit}" if v.unit else f"{v.value:g}"


def _length_mismatch(
    a: Optional[ValueUnit], b: Optional[ValueUnit], tolerance: float = 0.1
) -> bool:
    """True only when both sides are known, same unit, and differ by more
    than `tolerance` (catalog-rounding grace, same as `_shaft_compatible`).
    Missing data or a unit mismatch is NOT a claimed conflict — this
    module's exclude-on-missing rule applies to what gets reported, too.
    """
    if a is None or b is None:
        return False
    if a.value is None or b.value is None:
        return False
    if a.unit is None or b.unit is None or a.unit != b.unit:
        return False
    return abs(a.value - b.value) > tolerance


def mounting_conflicts(motor: Motor, gearhead: Gearhead) -> List[str]:
    """Human-readable mounting-geometry mismatches between a motor and a
    gearhead's input flange.

    Unlike `compatible_gearheads` (which filters silently), this reports
    WHY two specific parts don't fit — intended for a build/BOM UI
    surfacing explicit conflict reasons (see `todo/BUILD.md` Part 4's
    planned `_warnings` response field; not wired to the relations API
    yet — that's Build's job). Each string names the dimension and both
    sides' values, matching the style already used client-side in
    `app/frontend/src/utils/compat.ts`'s `detail` strings.

    Only compares a dimension when BOTH sides have it — missing data is
    never reported as a conflict, per this module's exclude-on-missing
    philosophy. An empty return means "no *known* conflict", which may
    mean compatible OR insufficient data; callers needing to
    distinguish those should check the relevant fields directly.

    Deliberately NOT checked (schema doesn't support it yet):
    - `shaft_length` vs. anything on Gearhead — `Gearhead` has no field
      for input-coupling engagement depth. `input_mounting_flange.
      pilot_depth` is a different physical dimension (flange register
      depth, for centering) and comparing shaft length against it would
      be a category error, not an approximation.
    - `shaft_modification` — `Gearhead` has no "required shaft
      modification" field to compare against.
    """
    conflicts: List[str] = []

    if motor.shaft_diameter is not None and gearhead.input_shaft_diameter is not None:
        if not _shaft_compatible(motor.shaft_diameter, gearhead.input_shaft_diameter):
            conflicts.append(
                f"shaft diameter: motor {_fmt(motor.shaft_diameter)} exceeds "
                f"gearhead input bore {_fmt(gearhead.input_shaft_diameter)}"
            )

    if (
        motor.motor_mount_pattern is not None
        and gearhead.input_motor_mount
        and motor.motor_mount_pattern not in gearhead.input_motor_mount
    ):
        conflicts.append(
            f"mount pattern: motor {motor.motor_mount_pattern!r} not in gearhead "
            f"input_motor_mount {sorted(gearhead.input_motor_mount)}"
        )

    mf = motor.mounting_flange
    gf = gearhead.input_mounting_flange
    if mf is not None and gf is not None:
        if (
            mf.mounting_standard is not None
            and gf.mounting_standard is not None
            and mf.mounting_standard != gf.mounting_standard
        ):
            conflicts.append(
                f"mounting standard: motor {mf.mounting_standard} vs gearhead "
                f"input {gf.mounting_standard}"
            )
        if _length_mismatch(mf.bolt_circle_diameter, gf.bolt_circle_diameter):
            conflicts.append(
                f"bolt circle diameter: motor {_fmt(mf.bolt_circle_diameter)} vs "
                f"gearhead input {_fmt(gf.bolt_circle_diameter)}"
            )
        if _length_mismatch(mf.bolt_hole_diameter, gf.bolt_hole_diameter):
            conflicts.append(
                f"bolt hole diameter: motor {_fmt(mf.bolt_hole_diameter)} vs "
                f"gearhead input {_fmt(gf.bolt_hole_diameter)}"
            )
        if (
            mf.bolt_hole_count is not None
            and gf.bolt_hole_count is not None
            and mf.bolt_hole_count != gf.bolt_hole_count
        ):
            conflicts.append(
                f"bolt hole count: motor {mf.bolt_hole_count} vs gearhead input "
                f"{gf.bolt_hole_count}"
            )
        # Register/spigot fit — treated as an equality check like bolt
        # circle diameter. Vendors differ on which side is male/female,
        # so this doesn't assume a shaft-into-bushing "motor <= bore"
        # relationship the way `_shaft_compatible` does.
        if _length_mismatch(mf.pilot_diameter, gf.pilot_diameter):
            conflicts.append(
                f"pilot diameter: motor {_fmt(mf.pilot_diameter)} vs gearhead "
                f"input {_fmt(gf.pilot_diameter)}"
            )

    return conflicts

File: app/backend/src/routes/relations.ts
/**
 * Relations route — device compatibility queries.
 * GET /api/v1/relations/motors-for-actuator?id=<uuid>&type=linear_actuator|electric_cylinder
 * GET /api/v1/relations/drives-for-motor?id=<uuid>
 * GET /api/v1/relations/gearheads-for-motor?id=<uuid>
 *
 * Mirrors the predicates in `specodex/relations.py` (SCHEMA Phase 3a).
 * Two implementations are intentional during the Express → Python backend
 * migration (todo/PYTHON_BACKEND.md). The Python module is the future
 * source of truth; this file goes away when Express does. Until then,
 * the response shape mirrors `/api/v1/search` so the frontend table can
 * render results without new code.
 */

import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { DynamoDBService } from '../db/dynamodb';
import config from '../config';

const router = Router();
const db = new DynamoDBService({ tableName: config.dynamodb.tableName });

// ---------------------------------------------------------------------------
// Local types — narrow shapes for the fields the predicates read. Phase 1
// fields (motor_mount_pattern, compatible_motor_mounts, input_motor_mount,
// output_motor_mount) are not yet in app/backend/src/types/models.ts; that
// hand-typed mirror retires with the Express deletion in PYTHON_BACKEND.md
// Phase 3, so adding them there now is wasted work.
// ---------------------------------------------------------------------------

type ValueUnit = { value?: number | null; unit?: string | null };
type MinMaxUnit = { min?: number | null; max?: number | null; unit?: string | null };

interface MotorRecord {
  product_id: string;
  product_type: 'motor';
  motor_mount_pattern?: string | null;
  rated_voltage?: MinMaxUnit | null;
  rated_current?: ValueUnit | null;
  rated_torque?: ValueUnit | null;
  rated_speed?: ValueUnit | null;
  shaft_diameter?: ValueUnit | null;
  encoder_feedback_support?: string | null;
  [key: string]: unknown;
}

interface DriveRecord {
  product_id: string;
  product_type: 'drive';
  input_voltage?: MinMaxUnit | null;
  rated_current?: ValueUnit | null;
  encoder_feedback_support?: string[] | null;
  [key: string]: unknown;
}

interface GearheadRecord {
  product_id: string;
  product_type: 'gearhead';
  input_motor_mount?: string[] | null;
  output_motor_mount?: string | null;
  input_shaft_diameter?: ValueUnit | null;
  [key: string]: unknown;
}

interface ActuatorRecord {
  product_id: string;
  product_type: 'linear_actuator' | 'electric_cylinder';
  // LinearActuator: list of accepted mounts.
  compatible_motor_mounts?: string[] | null;
  // ElectricCylinder: single integrated mount.
  motor_mount_pattern?: string | null;
  // Requirements-narrowing fields (compatible_actuators floors).
  stroke?: ValueUnit | null;
  max_push_force?: ValueUnit | null;
  max_linear_speed?: ValueUnit | null;
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Predicates — port of specodex/relations.py. "Exclude on missing data"
// philosophy: any None / null / missing field returns false. Recall loss
// from incomplete records is the right failure mode for compatibility
// queries.
// ---------------------------------------------------------------------------

function valueGte(a: ValueUnit | null | undefined, b: ValueUnit | null | undefined): boolean {
  if (!a || !b) return false;
  if (a.value == null || b.value == null) return false;
  if (!a.unit || !b.unit || a.unit !== b.unit) return false;
  return a.value >= b.value;
}

function rangeWithin(
  inner: MinMaxUnit | null | undefined,
  outer: MinMaxUnit | null | undefined,
): boolean {
  if (!inner || !outer) return false;
  if (!inner.unit || !outer.unit || inner.unit !== outer.unit) return false;
  if (inner.min != null && outer.min != null && inner.min < outer.min) return false;
  if (inner.max != null && outer.max != null && inner.max > outer.max) return false;
  if (inner.min == null && inner.max == null) return false;
  return true;
}

function shaftCompatible(
  motorShaft: ValueUnit | null | undefined,
  gearheadInput: ValueUnit | null | undefined,
): boolean {
  if (!motorShaft || !gearheadInput) return false;
  if (motorShaft.value == null || gearheadInput.value == null) return false;
  if (!motorShaft.unit || !gearheadInput.unit) return false;
  if (motorShaft.unit !== gearheadInput.unit) return false;
  return Math.abs(motorShaft.value - gearheadInput.value) <= 0.1;
}

function encoderIntersect(motor: MotorRecord, drive: DriveRecord): boolean {
  const motorProto = motor.encoder_feedback_support;
  const driveProtos = drive.encoder_feedback_support;
  if (!motorProto || !driveProtos || driveProtos.length === 0) return false;
  return driveProtos.includes(motorProto);
}

/**
 * Port of relations.py `_meets_floor`: present, in canonical `unit`,
 * and >= `floor`. A different unit means a different physical quantity
 * or an un-normalised field — excluded either way (precision over
 * recall).
 */
function meetsFloor(
  value: ValueUnit | null | undefined,
  floor: number,
  unit: string,
): boolean {
  if (!value || value.value == null || !value.unit) return false;
  if (value.unit !== unit) return false;
  return value.value >= floor;
}

/**
 * Distribution-position metadata attached to each actuator candidate
 * in the `/actuators` response. Drives todo/BUILD.md Part 3's
 * "8th most common stroke in catalogue" badge. Computed once per
 * request from the candidate set (not the full catalogue), so the
 * rank is meaningful relative to what passed the filter.
 *
 * `spec` is fixed to `stroke` for the first cut — Build's most
 * user-facing dimension, and the one the doc's example targets.
 * Future expansion can pick per-request (peak_force_rating, peak_velocity_rating)
 * via a query parameter; not needed for Phase 1.
 */
type DistributionPosition = {
  spec: 'stroke' | 'peak_force_rating' | 'peak_velocity_rating';
  rank: number;
  cluster_count: number;
};

/**
 * Group actuators by integer-millimetre stroke, rank the resulting
 * clusters by size (1 = most populous), and attach a
 * `_distribution_position` block to each candidate.
 *
 * Actuators missing the field or carrying it in a non-canonical unit
 * receive no badge — same precision-over-recall rule the predicates
 * follow. Ties in cluster size are broken by the spec value
 * (ascending) so a stable rank survives across calls.
 */
function attachStrokeDistributionPositions(
  candidates: ActuatorRecord[],
): Array<ActuatorRecord & { _distribution_position?: DistributionPosition }> {
  const bucketKey = (v: ValueUnit | null | undefined): number | null => {
    if (!v || v.value == null || v.unit !== 'mm') return null;
    return Math.round(v.value);
  };

  const counts = new Map<number, number>();
  for (const c of candidates) {
    const k = bucketKey(c.stroke);
    if (k == null) continue;
    counts.set(k, (counts.get(k) ?? 0) + 1);
  }

  // Sort clusters by size descending, tiebreak on the stroke value
  // ascending so the rank is deterministic across identical-size
  // clusters.
  const ranked = [...counts.entries()].sort((a, b) => {
    if (b[1] !== a[1]) return b[1] - a[1];
    return a[0] - b[0];
  });
  const rankByBucket = new Map<number, number>();
  ranked.forEach(([bucket], i) => rankByBucket.set(bucket, i + 1));

  return candidates.map(c => {
    const k = bucketKey(c.stroke);
    if (k == null) return c;
    const rank = rankByBucket.get(k);
    const cluster_count = counts.get(k);
    if (rank == null || cluster_count == null) return c;
    return {
      ...c,
      _distribution_position: { spec: 'stroke', rank, cluster_count },
    };
  });
}

/**
 * Port of relations.py `compatible_actuators` — the requirements-first
 * entry point for Build's slot-fill sequence (todo/BUILD.md Part 4).
 * Each floor is optional and independent; an unset floor applies no
 * constraint (Build's "blank = no constraint applied" rule). A set
 * floor excludes actuators missing the field or carrying it in a
 * non-canonical unit.
 */
function compatibleActuators(
  actuatorDb: ActuatorRecord[],
  floors: {
    minStrokeMm?: number;
    minPeakForceN?: number;
    minPeakVelocityMmS?: number;
  },
): ActuatorRecord[] {
  return actuatorDb.filter(a => {
    if (floors.minStrokeMm != null && !meetsFloor(a.stroke, floors.minStrokeMm, 'mm')) {
      return false;
    }
    if (
      floors.minPeakForceN != null &&
      !meetsFloor(a.max_push_force, floors.minPeakForceN, 'N')
    ) {
      return false;
    }
    if (
      floors.minPeakVelocityMmS != null &&
      !meetsFloor(a.max_linear_speed, floors.minPeakVelocityMmS, 'mm/s')
    ) {
      return false;
    }
    return true;
  });
}

function compatibleMotors(
  actuator: ActuatorRecord,
  motorDb: MotorRecord[],
): MotorRecord[] {
  let mounts: Set<string>;
  if (actuator.product_type === 'linear_actuator') {
    mounts = new Set(actuator.compatible_motor_mounts || []);
  } else {
    mounts = actuator.motor_mount_pattern
      ? new Set([actuator.motor_mount_pattern])
      : new Set();
  }
  if (mounts.size === 0) return [];
  return motorDb.filter(m => m.motor_mount_pattern && mounts.has(m.motor_mount_pattern));
}

function compatibleDrives(motor: MotorRecord, driveDb: DriveRecord[]): DriveRecord[] {
  if (!motor.rated_voltage || !motor.rated_current) return [];
  return driveDb.filter(d => {
    if (!rangeWithin(motor.rated_voltage, d.input_voltage)) return false;
    if (!valueGte(d.rated_current, motor.rated_current)) return false;
    if (!encoderIntersect(motor, d)) return false;
    return true;
  });
}

function compatibleGearheads(
  motor: MotorRecord,
  gearheadDb: GearheadRecord[],
): GearheadRecord[] {
  if (!motor.motor_mount_pattern) return [];
  return gearheadDb.filter(g => {
    if (!g.input_motor_mount || g.input_motor_mount.length === 0) return false;
    if (!g.input_motor_mount.includes(motor.motor_mount_pattern!)) return false;
    if (!shaftCompatible(motor.shaft_diameter, g.input_shaft_diameter)) return false;
    return true;
  });
}

// ---------------------------------------------------------------------------
// Routes
// ---------------------------------------------------------------------------

const ActuatorIdQuery = z.object({
  id: z.string().min(1),
  type: z.enum(['linear_actuator', 'electric_cylinder']),
});

// Floors arrive as query strings; coerce + reject negatives/NaN. All
// optional — Build's progressive-narrowing rule. `min_duty_cycle` and
// `orientation` are accepted for the BUILD.md ActuatorQuery contract
// but deliberately don't filter: duty cycle is a motor-thermal concept
// and orientation a downstream derating hint (see relations.py
// compatible_actuators docstring).
const ActuatorsQuery = z.object({
  min_stroke_mm: z.coerce.number().nonnegative().optional(),
  min_peak_force_n: z.coerce.number().nonnegative().optional(),
  min_peak_velocity_mm_s: z.coerce.number().nonnegative().optional(),
  min_duty_cycle: z.coerce.number().min(0).max(1).optional(),
  orientation: z.enum(['horizontal', 'vertical']).optional(),
});

const MotorIdQuery = z.object({
  id: z.string().min(1),
});

router.get('/actuators', async (req: Request, res: Response): Promise<void> => {
  const parsed = ActuatorsQuery.safeParse(req.query);
  if (!parsed.success) {
    res.status(400).json({
      success: false,
      error: 'Invalid query parameters',
      details: parsed.error.issues.map(i => `${i.path.join('.')}: ${i.message}`),
    });
    return;
  }
  const q = parsed.data;
  console.log(
    `[relations] actuators stroke>=${q.min_stroke_mm ?? '-'} force>=${q.min_peak_force_n ?? '-'} velocity>=${q.min_peak_velocity_mm_s ?? '-'}`,
  );

  const actuators = (await db.list('linear_actuator')) as unknown as ActuatorRecord[];
  const matches = compatibleActuators(actuators, {
    minStrokeMm: q.min_stroke_mm,
    minPeakForceN: q.min_peak_force_n,
    minPeakVelocityMmS: q.min_peak_velocity_mm_s,
  });
  const annotated = attachStrokeDistributionPositions(matches);
  res.json({
    success: true,
    data: annotated,
    count: annotated.length,
    total: actuators.length,
  });
});

router.get('/motors-for-actuator', async (req: Request, res: Response): Promise<void> => {
  const parsed = ActuatorIdQuery.safeParse(req.query);
  if (!parsed.success) {
    res.status(400).json({
      success: false,
      error: 'Invalid query parameters',
      details: parsed.error.issues.map(i => `${i.path.join('.')}: ${i.message}`),
    });
    return;
  }
  const { id, type } = parsed.data;

  const safeId = id.replace(/\r|\n/g, '');
  console.log(`[relations] motors-for-actuator id=${safeId} type=${type}`);

  const actuator = (await db.read(id, type)) as ActuatorRecord | null;
  if (!actuator) {
    res.status(404).json({ success: false, error: 'Actuator not found' });
    return;
  }
  const motors = (await db.list('motor')) as unknown as MotorRecord[];
  const matches = compatibleMotors(actuator, motors);
  res.json({ success: true, data: matches, count: matches.length });
});

router.get('/drives-for-motor', async (req: Request, res: Response): Promise<void> => {
  const parsed = MotorIdQuery.safeParse(req.query);
  if (!parsed.success) {
    res.status(400).json({
      success: false,
      error: 'Invalid query parameters',
      details: parsed.error.issues.map(i => `${i.path.join('.')}: ${i.message}`),
    });
    return;
  }
  const { id } = parsed.data;

  const safeId = id.replace(/\r|\n/g, '');
  console.log(`[relations] drives-for-motor id=${safeId}`);

  const motor = (await db.read(id, 'motor')) as MotorRecord | null;
  if (!motor) {
    res.status(404).json({ success: false, error: 'Motor not found' });
    return;
  }
  const drives = (await db.list('drive')) as unknown as DriveRecord[];
  const matches = compatibleDrives(motor, drives);
  res.json({ success: true, data: matches, count: matches.length });
});

router.get('/gearheads-for-motor', async (req: Request, res: Response): Promise<void> => {
  const parsed = MotorIdQuery.safeParse(req.query);
  if (!parsed.success) {
    res.status(400).json({
      success: false,
      error: 'Invalid query parameters',
      details: parsed.error.issues.map(i => `${i.path.join('.')}: ${i.message}`),
    });
    return;
  }
  const { id } = parsed.data;

  const safeId = id.replace(/\r|\n/g, '');
  console.log(`[relations] gearheads-for-motor id=${safeId}`);

  const motor = (await db.read(id, 'motor')) as MotorRecord | null;
  if (!motor) {
    res.status(404).json({ success: false, error: 'Motor not found' });
    return;
  }
  const gearheads = (await db.list('gearhead')) as unknown as GearheadRecord[];
  const matches = compatibleGearheads(motor, gearheads);
  res.json({ success: true, data: matches, count: matches.length });
});

// Internal predicates exported for jest tests; not part of the route surface.
export const _predicates = {
  valueGte,
  rangeWithin,
  shaftCompatible,
  encoderIntersect,
  meetsFloor,
  compatibleActuators,
  compatibleMotors,
  compatibleDrives,
  compatibleGearheads,
  attachStrokeDistributionPositions,
};

export default router;
